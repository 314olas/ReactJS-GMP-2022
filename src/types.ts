export interface IMovie {
	id?: number
	title: string
	tagline: string
	vote_average: number
	vote_count: number
	release_date: string
	poster_path: string
	overview: string
	budget: number
	revenue: number
	genres: string[]
	runtime: number
}

export interface IMoviesResponse {
	totalAmount: number;
    data: IMovie[];
    offset: number;
    limit: number;
}

export enum MovieActionEnum {
	Delete = 'Delete',
	Edit = 'Edit',
	Add = 'Add',
}

export interface IDropdownData {
	name: string
	value: string
}

export type IObjectKey = {
    [key: string]: any;
};

export interface ICommonFormField {
    typeField?: string,
    value: string | number | IDropdownData[],
    placeholder?: string,
    label?: string,
    name?: string,
    step?: number,
    multiply?: boolean,
    data?: IDropdownData[] | null
}

export interface IAddMovieForm {
    title: ICommonFormField,
    tagline: ICommonFormField,
    budget: ICommonFormField,
    genres: ICommonFormField,
    release_date: ICommonFormField,
    vote_count: ICommonFormField,
    poster_path: ICommonFormField,
    overview: ICommonFormField,
    vote_average: ICommonFormField,
    runtime: ICommonFormField,
    revenue: ICommonFormField;
};

export interface IUpdateFormField {
    name: keyof IAddMovieForm,
    value: string | number | IDropdownData[],
}

export interface IErrorResponse {
    error: {
        data?: any,
        message?: any
    }
}
