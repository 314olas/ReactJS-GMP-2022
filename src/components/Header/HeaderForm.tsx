import React, { useRef } from 'react'

export type ISubmitFC = (e: React.FormEvent, value: string) => void

export interface IHeaderFormProps {
	onSubmitForm: ISubmitFC
}

export default function HeaderForm({ onSubmitForm }: IHeaderFormProps) {
	const input = useRef(null)

	return (
		<form className='header__search' onSubmit={(e) => { onSubmitForm(e, input.current.value.trim()); input.current.value = '' }}>
			<h2 className='title'>FIND YOUR MOViE</h2>
			<div className='header__search-field'>
				<input
					ref={input}
					type='text'
					className='input'
					placeholder='What do you want to watch?'
				/>
				<button data-testid="searchBtn" className='button button-warning'>search</button>
			</div>
		</form>
	)
}
