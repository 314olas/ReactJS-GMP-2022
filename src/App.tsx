import React, { useState } from 'react'
import { store } from './store/index';
import ErrorBoundary from './components/ErrorBoundary'
import Footer from './components/Footer'
import Header from './components/Header/Header'
import Filter from './components/Filter/Filter'
import MovieList from './components/Movie/MovieList'

import { Provider } from 'react-redux';
import { useAppSelector } from './hooks/redux';

export interface IAppProps {}

export default function App(props: IAppProps) {

	return (
		<Provider store={store}>
			<Header />
			<main className='content'>
			<Filter />
				<ErrorBoundary>
					<MovieList />
				</ErrorBoundary>
			</main>
			<Footer />
		</Provider>

	)
}
