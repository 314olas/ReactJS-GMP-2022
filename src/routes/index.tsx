import React from 'react'
import { createBrowserRouter, Navigate } from 'react-router-dom'
import App from '../App'
import ErrorPage from '../components/Error'

export const routers = createBrowserRouter([
	{
		path: '/',
		element: <Navigate to="/search" />,
	},
	{
		path: '/search',
		element: <App />,
	},
	{
		path: '/search/:searchQuery',
		element: <App />,
	},
	{
		path: '*',
		element: <ErrorPage />
	}
],
	{
		basename: '/'
	}
)
