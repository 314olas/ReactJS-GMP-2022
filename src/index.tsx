import React from 'react'
import ReactDOM from 'react-dom/client'
import { RouterProvider } from 'react-router-dom'
import { routers } from './routes'
import './styles/style.scss'

const root = ReactDOM.createRoot(document.getElementById('app'))
root.render(
	<React.StrictMode>
		<RouterProvider router={routers} />
	</React.StrictMode>
)
