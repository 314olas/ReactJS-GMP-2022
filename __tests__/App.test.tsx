import React from 'react'
import { MemoryRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import { fireEvent, render, RenderResult } from '@testing-library/react'
import { act } from 'react-dom/test-utils'

import { moviesListData } from './moks/movies'

import { store } from '../src/store'
import App from '../src/App'
import HeaderForm from '../src/components/Header/HeaderForm';
import userEvent from '@testing-library/user-event'

jest.mock('../src/hooks/useGetMovies', () => ({
	useGetMovies: jest.fn(() => {
		return {
			movies: moviesListData
		}
	})
}));

const renderApp = (): RenderResult =>
	render(
		<Provider store={store}>
			<MemoryRouter>
				<App />
			</MemoryRouter>
		</Provider>
	)

let container: HTMLElement;
let modalContainer: HTMLElement;

describe('App page', () => {
	let tree: RenderResult;
	const submitMockFunction = jest.fn()
	container = document.createElement('div');
	container.id = 'app';
	modalContainer = document.createElement('div');
	modalContainer.id = 'modal';
	document.body.appendChild(container);
	document.body.appendChild(modalContainer);

	beforeEach(() => {
		global.window = window
		window.scroll = jest.fn()
		window.HTMLElement.prototype.scrollIntoView = jest.fn()
	});

	it('check App snapshot', async () => {
		await act(() => {
			tree = renderApp()
		})
		expect(tree.asFragment()).toMatchSnapshot()
	})

	it('HeaderForm onSubmitForm function was called', async () => {
		submitMockFunction.mockImplementation((e, value) => {

		})
		const { getByPlaceholderText, findByTestId } = render(<HeaderForm onSubmitForm={submitMockFunction} />)
		const searchInput = await getByPlaceholderText('What do you want to watch?') as HTMLInputElement
		const searchBtn = await findByTestId('searchBtn') as HTMLButtonElement

		await userEvent.click(searchInput)
		fireEvent.change(searchInput, { target: { value: 'Name for movie' } })
		await userEvent.click(searchBtn)
		expect(submitMockFunction).toHaveBeenCalledTimes(1)
	})
})

