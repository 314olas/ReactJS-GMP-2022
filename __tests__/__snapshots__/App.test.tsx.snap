// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App page check App snapshot 1`] = `
<DocumentFragment>
  <header
    class="header"
  >
    <img
      alt="header background"
      class="header__background"
      src="test-file-stub"
    />
    <div
      class="position-relative"
    >
      <div
        class="header__top"
      >
        <a
          class="header__logo"
          href="/"
        >
          <h1
            class="logo-text"
          >
            <strong>
              netflix
            </strong>
            roulette
          </h1>
        </a>
        <button
          class="button header__add-btn"
          type="button"
        >
          + Add movie
        </button>
      </div>
      <form
        class="header__search"
      >
        <h2
          class="title"
        >
          FIND YOUR MOViE
        </h2>
        <div
          class="header__search-field"
        >
          <input
            class="input"
            placeholder="What do you want to watch?"
            type="text"
          />
          <button
            class="button button-warning"
            data-testid="searchBtn"
          >
            search
          </button>
        </div>
      </form>
    </div>
  </header>
  <main
    class="content"
  >
    <section
      class="filter"
    >
      <ul
        class="filter__category-list"
      >
        <li
          class="item  active"
        >
          <a
            class="link"
            href="/"
          >
            all
          </a>
        </li>
        <li
          class="item  "
        >
          <a
            class="link"
            href="/"
          >
            Documentary
          </a>
        </li>
        <li
          class="item  "
        >
          <a
            class="link"
            href="/"
          >
            Comedy
          </a>
        </li>
        <li
          class="item  "
        >
          <a
            class="link"
            href="/"
          >
            Crime
          </a>
        </li>
        <li
          class="item  "
        >
          <a
            class="link"
            href="/"
          >
            Horror
          </a>
        </li>
        <li
          class="item  "
        >
          <a
            class="link"
            href="/"
          >
            Romance
          </a>
        </li>
        <li
          class="item  "
        >
          <a
            class="link"
            href="/"
          >
            Drama
          </a>
        </li>
        <li
          class="item  "
        >
          <a
            class="link"
            href="/"
          >
            Action
          </a>
        </li>
        <li
          class="item  "
        >
          <a
            class="link"
            href="/"
          >
            Science Fiction
          </a>
        </li>
        <li
          class="item  "
        >
          <a
            class="link"
            href="/"
          >
            Adventure
          </a>
        </li>
        <li
          class="item  "
        >
          <a
            class="link"
            href="/"
          >
            Fantasy
          </a>
        </li>
      </ul>
      <div
        class="filter__sort"
      >
        <span
          class="filter__sort-text"
        >
          Sort by
        </span>
        <div
          class="outside"
        >
          <div
            class="dropdown  "
          >
            <div
              class="dropdown__content"
            >
              <button
                aria-expanded="false"
                class="button dropdown__btn"
                id="dropdownMenuButton1"
                type="button"
              >
                select options
                <svg
                  class="dropdown__arrow"
                  fill="none"
                  height="7"
                  viewBox="0 0 10 7"
                  width="10"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M1.17339 0C0.319169 0 -0.141788 1.00184 0.413939 1.65057L4.23746 6.11398C4.63642 6.57971 5.35674 6.57992 5.75597 6.11442L9.58401 1.65101C10.1403 1.0024 9.67943 0 8.82494 0H1.17339Z"
                    fill-rule="evenodd"
                  />
                </svg>
              </button>
            </div>
            <ul
              aria-labelledby="dropdownMenuButton1"
              class="dropdown__menu "
            >
              <li>
                <button
                  class="dropdown__menu-button"
                  type="button"
                >
                  realease date
                </button>
              </li>
              <li>
                <button
                  class="dropdown__menu-button"
                  type="button"
                >
                  budget
                </button>
              </li>
              <li>
                <button
                  class="dropdown__menu-button"
                  type="button"
                >
                  vote count
                </button>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </section>
    <section
      class="container container--3 movie-card-container"
    >
      <article
        class="movie-card"
      >
        <div
          class="image-wrapper"
        >
          <button
            type="button"
          >
            <img
              alt="TITLE"
              src="test-file-stub"
            />
          </button>
          <div
            class="outside"
          >
            <div
              class="dropdown  movie-card__more"
            >
              <div
                class="dropdown__content"
              >
                <button
                  aria-expanded="false"
                  class="button dropdown__btn"
                  id="dropdownMenuButton1"
                  type="button"
                >
                  <span>
                    ...
                  </span>
                </button>
              </div>
              <ul
                aria-labelledby="dropdownMenuButton1"
                class="dropdown__menu right"
              >
                <li>
                  <button
                    class="dropdown__menu-button"
                    type="button"
                  >
                    Edit
                  </button>
                </li>
                <li>
                  <button
                    class="dropdown__menu-button"
                    type="button"
                  >
                    Delete
                  </button>
                </li>
              </ul>
            </div>
          </div>
        </div>
        <footer
          class="movie-card__footer"
        >
          <div
            class="text-content"
          >
            <h4
              class="movie-name"
            >
              TITLE
            </h4>
            <span
              class="movie-category"
            />
          </div>
          <span
            class="year-label"
          >
            06-29-2006
          </span>
        </footer>
      </article>
      <article
        class="movie-card"
      >
        <div
          class="image-wrapper"
        >
          <button
            type="button"
          >
            <img
              alt="Movie 2"
              src="test-file-stub"
            />
          </button>
          <div
            class="outside"
          >
            <div
              class="dropdown  movie-card__more"
            >
              <div
                class="dropdown__content"
              >
                <button
                  aria-expanded="false"
                  class="button dropdown__btn"
                  id="dropdownMenuButton1"
                  type="button"
                >
                  <span>
                    ...
                  </span>
                </button>
              </div>
              <ul
                aria-labelledby="dropdownMenuButton1"
                class="dropdown__menu right"
              >
                <li>
                  <button
                    class="dropdown__menu-button"
                    type="button"
                  >
                    Edit
                  </button>
                </li>
                <li>
                  <button
                    class="dropdown__menu-button"
                    type="button"
                  >
                    Delete
                  </button>
                </li>
              </ul>
            </div>
          </div>
        </div>
        <footer
          class="movie-card__footer"
        >
          <div
            class="text-content"
          >
            <h4
              class="movie-name"
            >
              Movie 2
            </h4>
            <span
              class="movie-category"
            >
              Drama Biography Music
            </span>
          </div>
          <span
            class="year-label"
          >
            06-29-2006
          </span>
        </footer>
      </article>
    </section>
  </main>
  <footer
    class="footer"
  >
    <a
      class="header__logo"
      href="/"
    >
      <span
        class="logo-text"
      >
        <strong>
          netflix
        </strong>
        roulette
      </span>
    </a>
  </footer>
</DocumentFragment>
`;
